src/TestableIO.System.IO.Abstractions.Wrappers/FileAclExtensions.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System.Runtime.Versioning;
+using System.Security.AccessControl;
+
+namespace System.IO.Abstractions
+{
+    /// <summary>
+    ///     ACL (access control list) extension methods for <see cref="IFile" />.
+    /// </summary>
+    public static class FileAclExtensions
+    {
+#if FEATURE_FILE_SYSTEM_ACL_EXTENSIONS
+        /// <inheritdoc cref="FileSystemAclExtensions.GetAccessControl(FileInfo)"/>
+#else
+        /// <inheritdoc cref="File.GetAccessControl(string)"/>
+#endif
+        [SupportedOSPlatform("windows")]
+        public static FileSecurity GetAccessControl(
+            this IFile file, string path)
+        {
+            IFileInfo fileInfo = file.FileSystem.FileInfo.New(path);
+            IFileSystemExtensibility extensibility =
+                fileInfo.Extensibility;
+            return extensibility.TryGetWrappedInstance(out FileInfo fi)
```

Comment:
I don't understand how you would solve this otherwise.
In order to implement the extension methods, I need access to the underlying instance (if it is available) and otherwise be able to store or retrieve metadata.
That's now just one extension method that I implemented in the TestableIO.System.IO.Abstractions.Wrappers project; but I need the same functionality, when I implement it in [Testably.Abstractions.AccessControl](https://github.com/Testably/Testably.Abstractions/tree/main/Source/Testably.Abstractions.AccessControl), so the interface itself needs to be public.

What I could do is to "remove" it from the `IDirectoryInfo` and "just add it as a separate interface to `MockDirectoryInfo` and `DirectoryInfoWrapper`. In the extension methods, I would then change it to 
```csharp
public static DirectorySecurity GetAccessControl(
            this IDirectoryInfo directoryInfo,
            AccessControlSections includeSections)
        {
            IFileSystemExtensibility extensibility =
                directoryInfo as IFileSystemExtensibility;
            return extensibility.TryGetWrappedInstance(out DirectoryInfo di)
                ? di.GetAccessControl(includeSections)
                : extensibility.RetrieveMetadata<DirectorySecurity>(
                    "AccessControl:DirectorySecurity") ?? new DirectorySecurity();
        }
```
Is this a better approach?

Reply:
Hm, what I don't like about this approach is the "broadness" of the interface `IFileSystemExtensibility`.

Could we make this more specific, e.g. like this:

```csharp
public static DirectorySecurity GetAccessControl(
            this IDirectoryInfo directoryInfo,
            AccessControlSections includeSections)
{
    return directoryInfo is IDirectoryInfoAclSupport directoryInfoAclSupport
        ? directoryInfoAclSupport.GetAccessControl(includeSections)
        : null
}
```

And for the write operations:


```csharp
public static DirectorySecurity SetAccessControl(
            this IDirectoryInfo directoryInfo,
            DirectorySecurity directorySecurity)
{
    return directoryInfo is IDirectoryInfoAclSupport directoryInfoAclSupport
        ? directoryInfoAclSupport.SetAccessControl(directorySecurity)
        : throw new NotSupportedException();
}
```

Your custom `Mock` classes could then easily implement whatever works best for `IDirectoryInfoAclSupport`.

Reply:
If I add explicit types to the interface, I still need the reference to `System.IO.FileSystem.AccessControl`, which is the whole point of this pull request :-(

I could do the following:
1. Remove the `Extensibility` properties from the interfaces
2. Add the extensibility interface instead on the corresponding wrapper or mock classes, e.g.
   ```csharp
       public class MockDirectoryInfo : DirectoryInfoBase, IFileSystemExtensibility
   ```
3. Implement the extensibility interface, e.g.
   ```csharp
        /// <inheritdoc cref="IFileSystemExtensibility.TryGetWrappedInstance{T}(out T)" />
        public bool TryGetWrappedInstance<T>(out T wrappedInstance)
            => GetExtensibility().TryGetWrappedInstance<T>(out wrappedInstance);

        /// <inheritdoc cref="IFileSystemExtensibility.StoreMetadata{T}(string, T)" />
        public void StoreMetadata<T>(string key, T value)
            => GetExtensibility().StoreMetadata<T>(key, value);

        /// <inheritdoc cref="IFileSystemExtensibility.RetrieveMetadata{T}(string)" />
        public T RetrieveMetadata<T>(string key)
        => GetExtensibility().RetrieveMetadata<T>(key);
        
        private IFileSystemExtensibility GetExtensibility()
        {
            var mockFileData = mockFileDataAccessor.GetFile(directoryPath);
            return mockFileData?.Extensibility ?? FileSystemExtensibility.GetNullObject(
                () => CommonExceptions.CouldNotFindPartOfPath(directoryPath));
        }
   ```
4. Adjust the ACL extension methods, e.g.
   ```csharp
        public static void Create(this IDirectoryInfo directoryInfo, DirectorySecurity directorySecurity)
        {
            IFileSystemExtensibility extensibility = directoryInfo as IFileSystemExtensibility;
            if (extensibility == null)
            {
                throw new NotSupportedException("The directory info does not support extensibility methods.")
            }

            if (extensibility.TryGetWrappedInstance(out DirectoryInfo di))
            {
                di.Create(directorySecurity);
            }
            else
            {
                directoryInfo.Create();
                extensibility.StoreMetadata("AccessControl:DirectorySecurity",
                    directorySecurity);
            }
        }
   ```

This way, the extensibility is "hidden" in normal use and more like an internal interface, even though it is public ;-)

Reply:
Ah sorry, I was back and forth with my example. I meant this:

```csharp

```csharp
public static object GetAccessControl(
            this IDirectoryInfo directoryInfo,
            object includeSections)
{
    return directoryInfo is IDirectoryInfoAclSupport directoryInfoAclSupport
        ? directoryInfoAclSupport.GetAccessControl(includeSections)
        : null
}
```

And for the write operations:


```csharp
public static void SetAccessControl(
            this IDirectoryInfo directoryInfo,
            object directorySecurity)
{
    if (directoryInfo is IDirectoryInfoAclSupport directoryInfoAclSupport)
        directoryInfoAclSupport.SetAccessControl(directorySecurity);
    else 
        throw new NotSupportedException();
}
```

Your suggestion is a little bit better, but still has the broad extensibility interface.

Reply:
I replaced the `IFileSystemExtensibility` with the following interface:
```csharp
    /// <summary>
    ///     Support ACL functionality on file system instances.
    /// </summary>
    public interface IFileSystemAclSupport
    {
        /// <summary>
        ///     Gets a access control object that encapsulates the access control list (ACL) entries for the file or directory in the file system.
        /// </summary>
        [SupportedOSPlatform("windows")]
        object GetAccessControl();

        /// <summary>
        ///     Gets a access control object that encapsulates the access control list (ACL) entries for the file or directory in the file system.
        /// </summary>
        /// <param name="includeSections">One of the <see cref="AccessControlSections" /> values that specifies the type of access control list (ACL) information to receive.</param>
        [SupportedOSPlatform("windows")]
        object GetAccessControl(AccessControlSections includeSections);

        /// <summary>
        ///     Applies access control list (ACL) entries described by the <paramref name="value"/> object to the file or directory in the file system.
        /// </summary>
        [SupportedOSPlatform("windows")]
        void SetAccessControl(object value);

        /// <summary>
        /// Specifies which sections of a security descriptor to save or load.</summary>
        [Flags]
        public enum AccessControlSections
        {
            /// <summary>
            /// No sections.
            /// </summary>
            None = 0,

            /// <summary>
            /// The system access control list (SACL).
            /// </summary>
            Audit = 1,

            /// <summary>
            /// The discretionary access control list (DACL).
            /// </summary>
            Access = 2,

            /// <summary>
            /// The owner.
            /// </summary>
            Owner = 4,

            /// <summary>
            /// The primary group.
            /// </summary>
            Group = 8,

            /// <summary>
            /// The entire security descriptor.
            /// </summary>
            All = Group | Owner | Access | Audit
        }
    }
```

The FileSystemExtensibility is now internal in the TestingHelpers project.

Is this, how you meant your suggestion, @fgreinacher ?

Reply:
Looks great now!